app/domain/authentication/handler/authentication_handler.rb

```diff
@@ -0,0 +1,155 @@
+      def initialize(
+        authenticator_type:,
+        role: ::Role,
+        resource: ::Resource,
+        authn_repo: DB::Repository::AuthenticatorRepository,
+        namespace_selector: Authentication::Util::NamespaceSelector,
+        logger: Rails.logger,
+        authentication_error: LogMessages::Authentication::AuthenticationError
+      )
+        @role = role
+        @resource = resource
+        @authenticator_type = authenticator_type
+        @logger = logger
+        @authentication_error = authentication_error
+
+        # Dynamically load authenticator specific classes
+        namespace = namespace_selector.select(
+          authenticator_type: authenticator_type
+        )
+
+        @identity_resolver = "#{namespace}::ResolveIdentity".constantize
+        @strategy = "#{namespace}::Strategy".constantize
+        @authn_repo = authn_repo.new(
+          data_object: "#{namespace}::DataObjects::Authenticator".constantize
+        )
+      end
+
+      def call(parameters:, request_ip:)
+        raise Errors::Authentication::RequestBody::MissingRequestParam, parameters[:code] unless parameters[:code]
+        raise Errors::Authentication::RequestBody::MissingRequestParam, parameters[:state] unless parameters[:state]
+        # Load Authenticator policy and values (validates data stored as variables)
+        authenticator = @authn_repo.find(
+          type: @authenticator_type,
+          account: parameters[:account],
+          service_id: parameters[:service_id]
+        )
+
+        if authenticator.nil?
+          raise(
+            Errors::Conjur::RequestedResourceNotFound,
+            "Unable to find authenticator with account: #{parameters[:account]} and service-id: #{parameters[:service_id]}"
+          )
+        end
+
+        role = @identity_resolver.new.call(
+          identity: @strategy.new(
+            authenticator: authenticator
+          ).callback(parameters),
+          account: parameters[:account],
+          allowed_roles: @role.that_can(
+            :authenticate,
+            @resource[authenticator.resource_id]
+          ).all
+        )
+
+        # TODO: Add an error message
+        raise 'failed to authenticate' unless role
+
+        unless role.valid_origin?(request_ip)
+          raise Errors::Authentication::InvalidOrigin
+        end
+
+        log_audit_success(authenticator, role, request_ip, @authenticator_type)
+
+        TokenFactory.new.signed_token(
+          account: parameters[:account],
+          username: role.role_id.split(':').last
+        )
+      rescue => e
+        log_audit_failure(parameters[:account], parameters[:service_id], request_ip, @authenticator_type, e)
+        handle_error(e)
+      end
+
+      def handle_error(err)
+        @logger.warn(@authentication_error.new(err))
+        case err
+        when Errors::Authentication::Security::RoleNotAuthorizedOnResource
+          puts "Errors::Authentication::Security::RoleNotAuthorizedOnResource"
```

Comment:
Out of curiosity why are we printing the error class here?

We could we have a `puts err.class` before the switch statement to eliminate most of the print statements